### PR TITLE
Lux profile

### PR DESCRIPTION
**Pull request by rmichaelis**
I'm not sure but maybe I should use a service instead of a directive. 

Please review.

### CODE REVIEW

**Comment by elemoine** [comment]
Thanks for the PR. I'll review this this afternoon.

**Comment by elemoine** [comment]
Why don't you use `ol.interaction.Draw` (in line string mode) instead of adding your own profile interaction?

File: geoportailv3/static/js/profile/profiledirective.js [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+  };
+
+  /**
+   * @type {Object}
+   */
+  var extractor = {z: z, dist: dist};
+
+  // Using closures for hoverCallback and outCallback since
+  // wrapping in angular.bind leads to a closure error.
+  // See PR https://github.com/google/closure-compiler/pull/867
```
Comment (elemoine) [inline geoportailv3/static/js/profile/profiledirective.js]: You wouldn't have that problem if you used `goog.bind`.

File: geoportailv3/templates/index.html [hunk 2]
```diff
@@ -114,8 +114,8 @@ <h2 translate>infos</h2>
       <div id="map" ngeo-map="mainCtrl.map">
         <app-infobar app-infobar-map="::mainCtrl.map"></app-infobar>
       </div>
+      <app-profile app-profiledata="mainCtrl.profiledata" app-profile-open="mainCtrl.profileOpen" app-profile-map="mainCtrl.map"></app-profile>
```
Comment (elemoine) [inline geoportailv3/templates/index.html]: You want to use one-timing binding fo `app-profile-map`.

File: geoportailv3/static/js/measure/measuredirective.js [hunk 3]
```diff
@@ -120,6 +154,39 @@ app.MeasureController = function($scope, ngeoDecorateInteraction) {
...
+        var req = $.param({
+          'geom': new ol.format.GeoJSON().writeGeometry(geom, encOpt),
+          'nbPoints': 100,
+          'layers': 'dhm'
+        });
+        var config = {
+          headers: {'Content-Type': 'application/x-www-form-urlencoded'}
+        };
+        this['profileOpen'] = true;
+        $http.post('profile.json', req, config).then(
```
Comment (elemoine) [inline geoportailv3/static/js/measure/measuredirective.js]: What's `profile.json` exactly?

Comment (rmichaelis) [inline geoportailv3/static/js/measure/measuredirective.js]: profile.json is the webservice that returns the profile for a selected line.

Comment (elemoine) [inline geoportailv3/static/js/measure/measuredirective.js]: So I guess we should use `appModule.constant` in the `index.html` template like for the other web services. See https://github.com/Geoportail-Luxembourg/geoportailv3/blob/master/geoportailv3/templates/index.html#L188-L193.

**Comment by rmichaelis** [comment]
> Why don't you use `ol.interaction.Draw` (in line string mode) instead of adding your own profile interaction?

@elemoine
There are essentially two reasons.
The first is that it seemed easier for me (perhaps wrongly) to interact with the event directly from interaction and thus to inherit all existing mechanisms provided by the base. Furthermore at first I wanted to extend `ngeo.interaction.MeasureLength`, because this interaction was doing almost all I wanted. But that was not possible.

The second is that other measurement tools were created in that way. Then I did in the same way, to keep a kind of consistency.

What would be the benefits of not using my own interaction? Performances?

**Comment by elemoine** [comment]
> @elemoine
> There are essentially two reasons.
> The first is that it seemed easier for me (perhaps wrongly) to interact with the event directly from interaction and thus to inherit all existing mechanisms provided by the base. Furthermore at first I wanted to extend ngeo.interaction.MeasureLength, because this interaction was doing almost all I wanted. But that was not possible.
> 
> The second is that other measurement tools were created in that way. Then I did in the same way, to keep a kind of consistency.
> 
> What would be the benefits of not using my own interaction? Performances?

I just thought that using an `ol.interaction.Draw` would be sufficient for the "profile" functionality. On the current http://map.geoportail.lu/ a simple line is drawn: there's no measurement values displayed as the user is drawing the line. Do you want a different behavior that the current one?

**Comment by elemoine** [comment]
> I just thought that using an ol.interaction.Draw would be sufficient for the "profile" functionality. On the current http://map.geoportail.lu/ a simple line is drawn: there's no measurement values displayed as the user is drawing the line. Do you want a different behavior that the current one?

And even if we want to display measurement values as the user is drawing the profile line on the map why not use `ngeo.interaction.MeasureLength` instead of adding a custom interaction to the geoportailv3 code base?

**Comment by rmichaelis** [comment]
> And even if we want to display measurement values as the user is drawing the profile line on the map why not use `ngeo.interaction.MeasureLength` instead of adding a custom interaction to the  geoportailv3 code base?


When I started the development, all my needs were not fulfilled with the version of  `ngeo.interaction.MeasureLength`. Now I'm able to use this interaction. 
But there is still a problem : the feature is not properly propagated with the `ngeo.MeasureEventType.MEASUREEND` event. 
I will open an issue for that.

**Comment by rmichaelis** [comment]
It will be ok as soon as #370 is fixed. 
Please review

File: geoportailv3/static/js/measure/measuredirective.js [hunk 4]
```diff
@@ -31,7 +32,9 @@ app.measureDirective = function(appMeasureTemplateUrl) {
     restrict: 'E',
     scope: {
       'map': '=appMeasureMap',
-      'active': '=appMeasureActive'
+      'active': '=appMeasureActive',
+      'profiledata': '=appProfiledata',
```
Comment (elemoine) [inline geoportailv3/static/js/measure/measuredirective.js]: Rename `appProfiledata` to `appMeasureProfiledata` and `appProfileOpen` to `appMeasureProfileopen` for consistency.

File: geoportailv3/static/js/measure/measuredirective.js [hunk 5]
```diff
@@ -120,6 +162,35 @@ app.MeasureController = function($scope, ngeoDecorateInteraction) {
   ngeoDecorateInteraction(measureAzimut);
   map.addInteraction(measureAzimut);
 
+  measureProfile.on(ngeo.MeasureEventType.MEASUREEND, goog.bind(function(evt) {
```
Comment (elemoine) [inline geoportailv3/static/js/measure/measuredirective.js]: We use `goog.events.listen` everywhere else. In this case:

```js
goog.events.listen(measureProfile, ngeo.MeasureEventType.MEASUREEND,
    /**
     * @param {ngeo.MeasureEvent} evt Measure event.
     */
    function(evt) {
      // …
    }, undefined, this);
```

File: geoportailv3/static/js/profile/profiledirective.js [hunk 6]
```diff
@@ -0,0 +1,237 @@
+/**
+ * @fileoverview This file provides a measure directive. This directive is used
+ * to create a measure panel in the page.
+ *
+ * Example:
+ *
+ * <app-measure app-measure-map="::mainCtrl.map"></app-measure>
```
Comment (elemoine) [inline geoportailv3/static/js/profile/profiledirective.js]: The example refers to the `app-measure` directive, which is not the directive this file includes.

File: geoportailv3/static/js/profile/profiledirective.js [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+/**
+ * @param {string} appProfileTemplateUrl Url to layermanager template
+ * @return {angular.Directive} The Directive Definition Object.
+ * @ngInject
+ */
+app.profileDirective = function(appProfileTemplateUrl) {
+  return {
+    restrict: 'E',
+    scope: {
+      'profiledata': '=appProfiledata',
```
Comment (elemoine) [inline geoportailv3/static/js/profile/profiledirective.js]: `appProfiledata` should be changed to `appProfileData`.

File: geoportailv3/static/js/profile/profiledirective.js [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+app.ProfileController = function($scope) {
+  /**
+   * @private
+   */
+  this.distanceLabel_ = 'Distance : ';
+
+  /**
+   * @private
+   */
+  this.elevationLabel_ = 'Elevation : ';
```
Comment (elemoine) [inline geoportailv3/static/js/profile/profiledirective.js]: Empty line missing after this line.

**Comment by elemoine** [comment]
I've just added a few minor comments, otherwise looks good to me.

**Comment by pgiraud** [comment]
@rmichaelis Can you please rebase this branch and fix the indentation issues?
For more chances to have travis happy you should run the check command.
If you don't have time for this, I can try to fix the indentation issues. Please tell me.

**Comment by rmichaelis** [comment]
May I merge this PR ?

File: geoportailv3/static/js/profile/profiledirective.js (hunk 1 above)
Comment (pgiraud) [inline geoportailv3/static/js/profile/profiledirective.js]: @rmichaelis have you tried using goog.bind as Éric suggested?

**Comment by pgiraud** [comment]
@rmichaelis did you see Éric' comment about using `goog.bind`?

**Comment by jaykayone** [comment]
@rmichaelis is on vacation until tomorrow

**Comment by pgiraud** [comment]
I propose we merge this for him and address remaining points later.

**Comment by jaykayone** [comment]
Ok

**Pull request merged**